building_sim_plugins/building_ignition_plugins/src/crowd_simulator.cpp

```diff
@@ -76,192 +80,167 @@ void CrowdSimulatorPlugin::PreUpdate(
+        _animate_idle_objects(delta_sim_time, ecm, crowd_sim_interface);
+      }
+      else
+      {
+        crowd_sim_interface->one_step_sim();
+        _update_all_objects(delta_sim_time, ecm, crowd_sim_interface);
       }
     }
   }
-  return true;
 }
 
 //==========================================================
 void CrowdSimulatorPlugin::_init_spawned_agents(
-  ignition::gazebo::EntityComponentManager& ecm)
+  ignition::gazebo::EntityComponentManager& ecm,
+  std::shared_ptr<crowd_simulator::CrowdSimInterface> crowd_sim_interface)
 {
   // check all the models are in the world
   std::unordered_map<std::string, size_t> objects_name;
-  size_t object_count = _crowd_sim_interface->get_num_objects();
-  for (size_t id = 0; id < object_count; id++)
+  size_t object_count = crowd_sim_interface->get_num_objects();
+  if (crowd_sim_interface->_menge_disabled)
   {
-    auto obj = _crowd_sim_interface->get_object_by_id(id);
-    // already found in the Dic
-    if (_entity_dic.find(obj->model_name) != _entity_dic.end())
-      continue;
-    objects_name.insert({obj->model_name, id});
-  }
-  // for external agent
-  ecm.Each<ignition::gazebo::components::Model,
-    ignition::gazebo::components::Name>(
-    [&](const ignition::gazebo::Entity& entity,
-    const ignition::gazebo::components::Model*,
-    const ignition::gazebo::components::Name* name) -> bool
+    for (size_t id = 0; id < object_count; id++)
     {
-      auto it_objects_name = objects_name.find(name->Data());
-      if (it_objects_name != objects_name.end())
+      objects_name.insert({crowd_sim_interface->get_internal_agent(id), id});
+    }
+    // for internal agent
+    ecm.Each<ignition::gazebo::components::Actor,
+      ignition::gazebo::components::Name>(
+      [&](const ignition::gazebo::Entity& entity,
+      const ignition::gazebo::components::Actor* actor,
+      const ignition::gazebo::components::Name* name) -> bool
       {
-        // update in entityDic
-        _entity_dic[name->Data()] = entity;
-        auto obj_ptr =
-        _crowd_sim_interface->get_object_by_id(it_objects_name->second);
-        // config internal spawned agent for custom trajectory
-        if (!obj_ptr->is_external)
+        auto it_objects_name = objects_name.find(name->Data());
+        if (it_objects_name != objects_name.end())
         {
-          _config_spawned_agents(obj_ptr, entity, ecm);
+          // update in entityDic
+          _entity_dic[name->Data()] = entity;
+          // config internal spawned agent for custom trajectory
+          _config_spawned_agents(nullptr, entity, ecm, crowd_sim_interface);
+          objects_name.erase(name->Data());
         }
-        objects_name.erase(name->Data());
-        RCLCPP_INFO(_crowd_sim_interface->logger(),
-        "Crowd Simulator found agent: " + name->Data() );
-      }
-      return true;
+        return true;
+      });
+    if (objects_name.size() != 0)
+    {
+      _initialized = false;
+      return;
```

Comment:
Fred initially compiled the objects_name list and removed the object one by one from the list if it is found, until the list become empty and deem it as initialized. I just keep it that way since it ain't broke.

Reply:
But the previous implementation spawned the actors one by one at runtime while the current one loads all of them at initialization before running the plugin so they are fundamentally different. I'd check if these cases happen at all and simplify the logic if they don't.

Reply:
This deal with part (2) from menge as discussed before.